### DISCUSSION TITLE
using podsync for people who don't know very well programming

### DISCUSSION THREAD

**Discussion by ahmedahmedo** [mxpv/podsync#262] [General]
1. get [YouTube API key](https://elfsight.com/blog/2016/12/how-to-get-youtube-api-key-tutorial/)
2. download and install [docker](https://www.docker.com/products/docker-desktop)
3. install [youtube-dl](https://ytdl-org.github.io/youtube-dl/index.html) for your operation system
3. create configuration file([config.toml](https://github.com/mxpv/podsync#configuration-example)) in default folder
  a. insert YouTube API key
  b. insert YouTube channel url
4. [pull and run](https://github.com/mxpv/podsync#run-via-docker) podsync files via docker 
5. open web browser and go http://localhost:8080, you will see xml files, copy link and paste your rss reader.

this tutor is very basic, please someone edit by means of programming and English

**Comment by elsiehupp** [comment]
Does Podsync strictly require Docker? If no, is there a reason I should use Docker? And if I'm not using Docker, how would I install Podsync without it?

I tried using `git` and `make,` and it failed on a ton of dependencies from GitHub; then I tried using Docker, and I didn't even know where to start...

How realistic would it be for me to be able to install Podsync from something like `apt`?

**Comment by elsiehupp** [comment]
Here's what I got when I tried to `make` the `git` clone:

> $ make
> go build -o bin/podsync ./cmd/podsync
> cmd/podsync/updater.go:13:2: cannot find package "github.com/hashicorp/go-multierror" in any of:
> 	/usr/lib/go-1.10/src/github.com/hashicorp/go-multierror (from $GOROOT)
> 	/home/rss/go/src/github.com/hashicorp/go-multierror (from $GOPATH)
> cmd/podsync/main.go:12:2: cannot find package "github.com/jessevdk/go-flags" in any of:
> 	/usr/lib/go-1.10/src/github.com/jessevdk/go-flags (from $GOROOT)
> 	/home/rss/go/src/github.com/jessevdk/go-flags (from $GOPATH)
> cmd/podsync/updater.go:17:2: cannot find package "github.com/mxpv/podsync/pkg/builder" in any of:
> 	/usr/lib/go-1.10/src/github.com/mxpv/podsync/pkg/builder (from $GOROOT)
> 	/home/rss/go/src/github.com/mxpv/podsync/pkg/builder (from $GOPATH)
> cmd/podsync/main.go:17:2: cannot find package "github.com/mxpv/podsync/pkg/config" in any of:
> 	/usr/lib/go-1.10/src/github.com/mxpv/podsync/pkg/config (from $GOROOT)
> 	/home/rss/go/src/github.com/mxpv/podsync/pkg/config (from $GOPATH)
> cmd/podsync/main.go:18:2: cannot find package "github.com/mxpv/podsync/pkg/db" in any of:
> 	/usr/lib/go-1.10/src/github.com/mxpv/podsync/pkg/db (from $GOROOT)
> 	/home/rss/go/src/github.com/mxpv/podsync/pkg/db (from $GOPATH)
> cmd/podsync/updater.go:20:2: cannot find package "github.com/mxpv/podsync/pkg/feed" in any of:
> 	/usr/lib/go-1.10/src/github.com/mxpv/podsync/pkg/feed (from $GOROOT)
> 	/home/rss/go/src/github.com/mxpv/podsync/pkg/feed (from $GOPATH)
> cmd/podsync/main.go:19:2: cannot find package "github.com/mxpv/podsync/pkg/fs" in any of:
> 	/usr/lib/go-1.10/src/github.com/mxpv/podsync/pkg/fs (from $GOROOT)
> 	/home/rss/go/src/github.com/mxpv/podsync/pkg/fs (from $GOPATH)
> cmd/podsync/updater.go:22:2: cannot find package "github.com/mxpv/podsync/pkg/model" in any of:
> 	/usr/lib/go-1.10/src/github.com/mxpv/podsync/pkg/model (from $GOROOT)
> 	/home/rss/go/src/github.com/mxpv/podsync/pkg/model (from $GOPATH)
> cmd/podsync/main.go:20:2: cannot find package "github.com/mxpv/podsync/pkg/ytdl" in any of:
> 	/usr/lib/go-1.10/src/github.com/mxpv/podsync/pkg/ytdl (from $GOROOT)
> 	/home/rss/go/src/github.com/mxpv/podsync/pkg/ytdl (from $GOPATH)
> cmd/podsync/updater.go:14:2: cannot find package "github.com/pkg/errors" in any of:
> 	/usr/lib/go-1.10/src/github.com/pkg/errors (from $GOROOT)
> 	/home/rss/go/src/github.com/pkg/errors (from $GOPATH)
> cmd/podsync/main.go:13:2: cannot find package "github.com/robfig/cron/v3" in any of:
> 	/usr/lib/go-1.10/src/github.com/robfig/cron/v3 (from $GOROOT)
> 	/home/rss/go/src/github.com/robfig/cron/v3 (from $GOPATH)
> cmd/podsync/main.go:14:2: cannot find package "github.com/sirupsen/logrus" in any of:
> 	/usr/lib/go-1.10/src/github.com/sirupsen/logrus (from $GOROOT)
> 	/home/rss/go/src/github.com/sirupsen/logrus (from $GOPATH)
> cmd/podsync/main.go:15:2: cannot find package "golang.org/x/sync/errgroup" in any of:
> 	/usr/lib/go-1.10/src/golang.org/x/sync/errgroup (from $GOROOT)
> 	/home/rss/go/src/golang.org/x/sync/errgroup (from $GOPATH)
> cmd/podsync/main.go:22:2: cannot find package "gopkg.in/natefinch/lumberjack.v2" in any of:
> 	/usr/lib/go-1.10/src/gopkg.in/natefinch/lumberjack.v2 (from $GOROOT)
> 	/home/rss/go/src/gopkg.in/natefinch/lumberjack.v2 (from $GOPATH)
> Makefile:12: recipe for target 'build' failed
> make: *** [build] Error 1

Well, I went ahead and painstakingly downloaded each and every one of the dependencies with `go get`, then tried `make` again only to get this:

> $ make
> go build -o bin/podsync ./cmd/podsync
> \# github.com/hashicorp/go-multierror
> ../go/src/github.com/hashicorp/go-multierror/multierror.go:112:9: undefined: errors.As
> ../go/src/github.com/hashicorp/go-multierror/multierror.go:117:9: undefined: errors.Is
> Makefile:12: recipe for target 'build' failed
> make: *** [build] Error 2

Any thoughts on where I should go from here?

**Comment by elsiehupp** [comment]
(Note that I have no idea what I'm doing when it comes to `go`.)

**Comment by ahmedahmedo** [comment]
this procedure actually for Mac OS user, I don't know your operation system but you can apply this procedure your operation system too. 
I think firstly you should search how to install docker and how to install any app into the docker.
docker likes operation system. 
first install docker after that install podsync into the docker.

maybe @mxpv can prepare elaborate how to guide

**Comment by s39n** [comment]
> (Note that I have no idea what I'm doing when it comes to `go`.)

What operating system are you using.  I have mine running on windows. But have got Linux in the past so I can help

**Comment by elsiehupp** [comment]
> What operating system are you using. I have mine running on windows. But have got Linux in the past so I can help

I'm using Ubuntu 18.04 LTS. I'm trying again with Docker—the problem I had before seems to have been that the user I was logged in with was not part of the `docker` group—and I'll see if I can get it to work this time. It would still be nice to have proper instructions for installing outside of Docker even if I get Docker to work, though.

**Comment by elsiehupp** [comment]
I have Podsync (apparently) running now in Docker on Ubuntu, and I have a few questions:

1. Where is the correct location for the config.toml file? Why does the install seemingly create a directory with that name rather than a file?
2. How do I test my first feed on a headless server, i.e. not localhost?
3. What does "404 page not found" (not the regular Apache 404 page) in my web browser mean? Is this a firewall issue?

For the time being you can reference these URLs:
http://rss.elsiehupp.com (Apache virtual host on port 80)
https://rss.elsiehupp.com (Apache virtual host on port 443)
http://rss.elsiehupp.com:8080 (blank page)
http://rss.elsiehupp.com:8080/ID1.xml ("404 page not found")

4. Is it possible to get Podsync RSS feeds running on port 80 or 443 alongside (or inside) Apache?

**Comment by s39n** [comment]
> > What operating system are you using. I have mine running on windows. But have got Linux in the past so I can help
> 
> I'm using Ubuntu 18.04 LTS. I'm trying again with Docker—the problem I had before seems to have been that the user I was logged in with was not part of the `docker` group—and I'll see if I can get it to work this time. It would still be nice to have proper instructions for installing outside of Docker even if I get Docker to work, though.

Ya that would definitely be an issue. I think the instructions are not to detailed that way you can install it on any operating system.

**Comment by elsiehupp** [comment]
To reproduce where I am right now, follow these steps on an Ubuntu 18.04 x64 system:
### Install prerequisites from `apt`
1. `sudo apt-get install ffmpeg`
2. `sudo apt-get install youtube-dl`
3. `sudo apt-get install make`
### Set up correct version of `go`
4. ~~`sudo apt-get install golang-go`~~ !! Installs wrong version of Go. Podsync requires 1.13.6.
 `wget https://dl.google.com/go/go1.13.6.linux-amd64.tar.gz`
5. `sudo tar -C /usr/local -xzf go1.13.6.linux-amd64.tar.gz`
6. `nano ~/.profile`
 at end of file add:
 `export GOROOT=/usr/local/go`
 `export PATH=$GOPATH/bin:$GOROOT/bin:$PATH`
 Save and exit.
 (This could probably be done better with `echo "something" >> ~./profile`, but I haven't tested it.)
### Install `podsync` from source
7. `git clone https://github.com/mxpv/podsync`
8. `nano ./podsync/Makefile`
9. After `build:` add a new line with `go get ./...` (to dynamically include dependencies), save and exit.
10. `cd ./podsync`
11. `make`
12. `go test -v ./…` (should run diagnostic tests)
### Set up and run `podsync`
13. `nano ./bin/config.toml`
14. Paste sample `config.toml`from `README.md`, save and exit.
15. To run `podsync` run `~/podsync/bin/podsync` (It appears the contents of the `bin` folder can be placed wherever you want them.)

At this point I am running into several problems:
1. The `podsync.log` file is formatted for `cat`, with strings like `^[[31m`, `^[[36m`, and `^[[0m` that don't work in editors like `nano`. These should probably be stripped out.
2. Auto-updating `youtube-dl` doesn't seem to work when it has been installed from `apt-get`. I feel like this should be more explicit.
> INFO[2020-06-22T23:04:08Z] updating youtube-dl                          
ERRO[2020-06-22T23:04:09Z] [debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--update', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.03.14
[debug] Python version 3.6.9 (CPython) - Linux-4.15.0-106-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.6-0ubuntu0.18.04.1, ffprobe 3.4.6-0ubuntu0.18.04.1, phantomjs ., rtmpdump 2.4
[debug] Proxy map: {}
Usage: youtube-dl [OPTIONS] URL [URL...]
>
>youtube-dl: error: youtube-dl's self-update mechanism is disabled on Debian.
Please update youtube-dl using apt(8).
See https://packages.debian.org/sid/youtube-dl for the latest packaged version.
  error="failed to execute youtube-dl: exit status 2"
ERRO[2020-06-22T23:04:09Z] failed to update youtube-dl                   error="failed to self update youtube-dl: failed to execute youtube-dl: exit status 2"
3. I'm getting the following error when trying to update a feed:
>ERRO[2020-06-22T23:11:18Z] failed to update feed: https://www.youtube.com/user/LinusTechTips  error="update failed: failed to query channel: context canceled"

 Is the feed URL properly formatted? I tried using a YouTube RSS feed, and `podsync` didn't seem to like this, either.
### Thoughts
I don't know about all y'all, but the fact that the installation requires a specific version of `go`, and I had to find this out through trial and error is kind of a problem.

The missing dependencies call in the `Makefile` might be a bug, or it might be an artifact of the authors of `podsync` intending for it to be built in a different environment, perhaps using `go get github.com/mxpv/podsync` instead of `git` and `make`.

Similarly, if `podsync` is going to use its own update mechanism for `youtube-dl` it should be smarter about not conflicting with `apt`.

For initial setup it would be ideal for there to be a preexisting sample `config.toml` with *real values* so that the user can confirm that the setup is, in fact, working. There is a test script `go test -v ./…` that uses the source, so perhaps that could be reproduced as a shell script for testing the binary. (I would kind of prefer using a shell script rather than a configuration file for production, and it would be nice to have fully documented command-line settings.)

Finally, this is still not working for me. Any thoughts of how I could resolve the `failed to query channel` error? Once I get the YouTube query working hopefully the only step remaining will be to get the RSS feeds running and at the URLs I want. (I'll probably think of or figure more things out on my own, but this is where I am at this exact point in time.)

**Comment by elsiehupp** [comment]
Trying to load `podsync` using `go`...
>**~$** go get github.com/mxpv/podsync
can't load package: package github.com/mxpv/podsync: no Go files in /usr/local/go/src/github.com/mxpv/podsync

Trying to load `podsync` in the `go` `src` directory:
>**/usr/local/go/src$** go get github.com/mxpv/podsync
go: finding github.com/mxpv/podsync latest
go: downloading github.com/mxpv/podsync v0.0.0-20200609053654-00c7bce6dc69
go: extracting github.com/mxpv/podsync v0.0.0-20200609053654-00c7bce6dc69
go: writing go.sum: open /usr/local/go/src/go.sum87694162.tmp: permission denied

Trying to build `podsync` in the `go` `src` directory rather than in `~`:
>**/usr/local/go/src/github.com/mxpv/podsync$** make
go build -o bin/podsync ./cmd/podsync
go: inconsistent vendoring in /usr/local/go/src/github.com/mxpv/podsync:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;go.mod requires github.com/mxpv/podsync  but vendor/modules.txt does not include it.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;run 'go mod tidy; go mod vendor' to sync
Makefile:12: recipe for target 'build' failed
make: *** [build] Error 1

I tried `go mod tidy; go mod vendor`, and it didn't make any difference in terms of the above error.

**Comment by elsiehupp** [comment]
I did some fussing, and it seems like there were two problems with my above setup:
1. `youtube-dl` is constantly being updated, and older versions are generally broken. `apt-get` installs a version of `youtube-dl` from ***2018***, which obviously no longer works.
2. `youtube-dl` is rather fussy with URLs. In my limited experimentation, it seems to work well with URLs formatted as `https://www.youtube.com/playlist?list=XXXXXXX`.

I was able to get `podsync` up and running, but I have a few remaining issues:
1. The built-in update mechanism for `youtube-dl` just... doesn't seem to work (because I'm not using Docker?). It appears I will need to set up `youtube-dl` updates with `cron`, I guess? I could migrate my `podsync` scheduling to `cron`, as well, just to keep everything in one place. I installed the latest `youtube-dl` using `curl`, and `podsync.log` included the following message:
>It looks like you installed youtube-dl with a package manager, pip, setup.py or a tarball. Please use that to update.
2. The way I'm using `podsync` is that it's just dumping the RSS feeds and episode files in an Apache server, so I would rather not be running `podsync`'s built-in server. It seems like I can ignore it, but `podsync` seems designed to run in the background as a daemon, so I will have to see how migrating to `cron` affects that.
3. Executing the `podsync` binary locks up the Terminal, even if I try to run it in the background with `podsync &`. This may not be a problem if I run `podsync` via `cron`, and closing the terminal window doesn't seem to interrupt the `podsync` process, but it's still rather annoying.
4. There isn't an obvious place to put the `podsync` executable, especially since it requires the `config.toml` file to be in the same directory and proceeds to dump `podsync.log` in the same directory, as well. It would be nice if `podsync` could be installed in `/usr/bin/` like a normal application and take command line arguments from a shell script, like `$podsync --config=PATH_TO_TOML --log=PATH_TO_LOG`, ya know?

I will try and follow up with updated Ubuntu installation instructions once I'm able to fiddle some more.

**Comment by bcarney66** [comment]
@ahmedahmedo:

> 4. create configuration file([config.toml](https://github.com/mxpv/podsync#configuration-example)) in default folder

I've followed these steps, but can't figure out where the "default folder" is on OS X.  Can you point out where it should go? Thank you.

**Comment by elsiehupp** [comment]
> I've followed these steps, but can't figure out where the "default folder" is on OS X. Can you point out where it should go?

You could try `~/Library/Containers/com.docker.docker/Data/`. There might be subfolders there, too. I don't use docker, so I don't know what would be in that folder.

**Comment by bcarney66** [comment]
Thanks, @elsiehupp. I appreciate the hint.

I tried putting the config.toml file in there, and I get the error message "**level=fatal msg="failed to load configuration file" error="failed to read config file: config.toml: read config.toml: is a directory"**" which is strange because it's not a directory.

I even scanned my entire hard drive to see if there is a directory someplace with that name, and couldn't find it.  This is kind of frustrating.

**Comment by elsiehupp** [comment]
Again, I'm kind of going blind, here—my previous answer was a guess via Google—but you could try moving `config.toml` up one level, i.e. to `~/Library/Containers/com.docker.docker/`. Apparently the "is a directory" error is the output of a specific Python function, and it may not literally mean what it says: [[link](https://stackoverflow.com/questions/52338706/isadirectoryerror-errno-21-is-a-directory-it-is-a-file)].

If that doesn't work, could you tell me what all appears in `~/Library/Containers/com.docker.docker/` and `~/Library/Containers/com.docker.docker/Data`? I could try and make another educated guess...

And once we do figure this out, I can try and get it added to the main setup instructions!

**Comment by bcarney66** [comment]
I moved it to `~/Library/Containers/com.docker.docker/` and that didn't work either. Inside `~/Library/Containers/com.docker.docker/` is just the `Data` directory.  Inside that is:

```
default.proraw
log (subdirectory)
task.lock
tasks (subdirectory)
vms (subdirectory)
```

In com.docker.backend.log there are a couple of mentions of config.toml, but they don't make sense to me.

```
time="2021-01-22T16:06:37-05:00" level=info msg="watching path /config.toml/"
time="2021-01-22T16:06:37-05:00" level=info msg="invalidating caches for /config.toml/"
time="2021-01-22T16:06:37-05:00" level=warning msg="unable to invalidate /config.toml/: ENOENT (this is probably safe)"
time="2021-01-22T16:06:37-05:00" level=info msg="watching path /data/"
time="2021-01-22T16:06:37-05:00" level=info msg="invalidating caches for /data/"
time="2021-01-22T16:06:37-05:00" level=warning msg="unable to invalidate /data/: ENOENT (this is probably safe)"
time="2021-01-22T16:06:37-05:00" level=info msg="grpcfuseClient.Add(0cfab10e23c1908c5150f3aec053f27b0c846f252486e5bdeed51ed62d9ca5ea, /exciting_haslett, [host=/data,VM=/data,dst=/app/data/,option= host=/config.toml,VM=/config.toml,dst=/app/config.toml,option=])"
time="2021-01-22T16:06:37-05:00" level=info msg="grpcfuseClient.Remove(0cfab10e23c1908c5150f3aec053f27b0c846f252486e5bdeed51ed62d9ca5ea)"
time="2021-01-22T16:06:37-05:00" level=info msg="unwatching path /config.toml/"
time="2021-01-22T16:06:37-05:00" level=info msg="vpnkitExposer.Add(0cfab10e23c1908c5150f3aec053f27b0c846f252486e5bdeed51ed62d9ca5ea, /exciting_haslett, [])"
time="2021-01-22T16:06:37-05:00" level=info msg="vpnkitExposer.Remove(0cfab10e23c1908c5150f3aec053f27b0c846f252486e5bdeed51ed62d9ca5ea)"
time="2021-01-22T16:06:37-05:00" level=info msg="invalidating caches for /config.toml/"
time="2021-01-22T16:06:37-05:00" level=warning msg="unable to invalidate /config.toml/: ENOENT (this is probably safe)"
```

**Comment by elsiehupp** [comment]
Okay, I'm back at home. Here's my process. First, these are the `podsync` launch commands:
```
$ docker pull mxpv/podsync:latest
$ docker run \
    -p 8080:8080 \
    -v $(pwd)/data:/app/data/ \
    -v $(pwd)/config.toml:/app/config.toml \
    mxpv/podsync:latest
```
`$(pwd)` seems like a shell variable, so we can do the following:
```
% echo $(pwd)
/Users/elsiehupp
```
Oh. Well, based on that I'm gonna try putting the sample `config.toml` at `~/config.toml`.
```
% docker run \
    -p 8080:8080 \
    -v $(pwd)/data:/app/data/ \
    -v $(pwd)/config.toml:/app/config.toml \
    mxpv/podsync:latest
time="2021-01-23T00:44:03Z" level=info msg="\n _______  _______  ______   _______           _        _______ \n(  ____ )(  ___  )(  __  \\ (  ____ \\|\\     /|( (    /|(  ____ \\\n| (    )|| (   ) || (  \\  )| (    \\/( \\   / )|  \\  ( || (    \\/\n| (____)|| |   | || |   ) || (_____  \\ (_) / |   \\ | || |      \n|  _____)| |   | || |   | |(_____  )  \\   /  | (\\ \\) || |      \n| (      | |   | || |   ) |      ) |   ) (   | | \\   || |      \n| )      | (___) || (__/  )/\\____) |   | |   | )  \\  || (____/\\\n|/       (_______)(______/ \\_______)   \\_/   |/    )_)(_______/\n"
time="2021-01-23T00:44:03Z" level=fatal msg="failed to load configuration file" error="failed to unmarshal toml: line 12: invalid TOML syntax"
```
Well, that's a different error! (I commented out the line in question, rinse and repeat, until there were no errors. It does appear to be loading `~/config.toml`.)

On further inspection, `pwd` is "[**p**rint **w**orking **d**irectory](https://www.cyberciti.biz/faq/pwd-linux-unix-command-examples/)". That is, it's wherever you run the command, which is by default `~`.

If you want to keep things tidy, you could create the following directories:
```
% mkdir ~/Docker
% mkdir ~/Docker/podsync
% mv ~/config.toml ~/Docker/podsync/config.toml
```
Then run `podsync` with the following command:
```
% docker run \
    -p 8080:8080 \
    -v ~/Docker/podsync/data:/app/data/ \
    -v ~/Docker/podsync/config.toml:/app/config.toml \
    mxpv/podsync:latest
```

**Comment by elsiehupp** [comment]
By the way, it does seem that if `config.toml` doesn't exist, `podsync` will create a directory called `config.toml`, which is why you got the "is a directory" error. I googled `ENOENT`, and it just means the file doesn't exist.

**Comment by bcarney66** [comment]
Thank you, @elsiehupp. I used the "tidy" method, and it does appear that it's using config.toml, as I had to comment out these lines:

```
# Bind a specific IP addresses for server ,"*": bind all IP addresses which is default option, localhost or 127.0.0.1  bind a single IPv4 address
# bind_address = "*" 
# Specify path for reverse proxy and only [A-Za-z0-9]
#path = "test"
```

Once I did that it seems to run. But when I go to localhost:8080, I just get a blank page. Looking at the source of the page it just says 

```
<pre>
</pre>
```

I commented out the log section so I could see any errors.  This is the last line that appears:

`time="2021-01-23T17:03:47Z" level=error msg="failed to update feed: https://www.youtube.com/channel/UC0qb5AAbSI9kcjptctjWs1g" error="update failed: failed to query channel: googleapi: Error 403: The request cannot be completed because you have exceeded your <a href=\"/youtube/v3/getting-started#quota\">quota</a>., quotaExceeded"`

I don't understand that line at all as this is the first time I've actually been able to connect. I know I created the access token and added it to the [tokens] section properly.

**Comment by bcarney66** [comment]
So I created a new API key, and now it appears to be working... It's downloading videos. I probably should have tested this on a channel with less hour long videos. :)

I appreciate your help, @elsiehupp !!!

**Comment by surfjz117** [comment]
I'm quite confused right now. 
My setup is a raspberry pi - running docker and portainer - but i can't deploy the image. 
I'm using tdeutsch/podsync:latest from docker.io because this is supporting arm of the raspberry.
I got the message config.toml is a directory. deletet the directory and put the config.toml at the same place and now getting no such file or directory. tryied to change the right to 777 but still getting the same error. 

any ideas?

**Reply by elsiehupp** [reply]
For the diagnosis, could you provide as much of the output as possible? You can put it between two sets of three “tick” marks, like so:

````
```
<pasted output>
```
````

**Reply by surfjz117** [reply]
INFO[2022-01-19T14:17:25Z] 

 _______  _______  ______   _______           _        _______ 

(  ____ )(  ___  )(  __  \ (  ____ \|\     /|( (    /|(  ____ \

| (    )|| (   ) || (  \  )| (    \/( \   / )|  \  ( || (    \/

| (____)|| |   | || |   ) || (_____  \ (_) / |   \ | || |      

|  _____)| |   | || |   | |(_____  )  \   /  | (\ \) || |      

| (      | |   | || |   ) |      ) |   ) (   | | \   || |      

| )      | (___) || (__/  )/\____) |   | |   | )  \  || (____/\

|/       (_______)(______/ \_______)   \_/   |/    )_)(_______/ 

FATA[2022-01-19T14:17:25Z] failed to load configuration file             error="failed to read config file: config.toml: open config.toml: no such file or directory"

INFO[2022-01-19T14:19:45Z] 

 _______  _______  ______   _______           _        _______ 

(  ____ )(  ___  )(  __  \ (  ____ \|\     /|( (    /|(  ____ \

| (    )|| (   ) || (  \  )| (    \/( \   / )|  \  ( || (    \/

| (____)|| |   | || |   ) || (_____  \ (_) / |   \ | || |      

|  _____)| |   | || |   | |(_____  )  \   /  | (\ \) || |      

| (      | |   | || |   ) |      ) |   ) (   | | \   || |      

| )      | (___) || (__/  )/\____) |   | |   | )  \  || (____/\

|/       (_______)(______/ \_______)   \_/   |/    )_)(_______/ 

FATA[2022-01-19T14:19:45Z] failed to load configuration file             error="failed to read config file: config.toml: open config.toml: no such file or directory"

**Reply by elsiehupp** [reply]
> I got the message config.toml is a directory. deletet the directory and put the config.toml at the same place and now getting no such file or directory. tryied to change the right to 777 but still getting the same error.

It *sounds* like you did the correct thing here, but the new error makes it sound like maybe your `config.toml` is still in the wrong location relative to the Docker container’s working directory? I mean, `config.toml` not being in the correct location is the simplest interpretation of the second error, `[...] config.toml: no such file or directory`.

What command are you using to run the Docker container? If you use the following:

```bash
$ docker run \
    -p 8080:8080 \
    -v $(pwd)/data:/app/data/ \
    -v $(pwd)/config.toml:/app/config.toml \
    tdeutsch/podsync:latest
```

`$(pwd)` is simply the directory where you are running the command (`pwd` being short for `print working directory`). If you use this command, `podsync` will look for `config.toml` at `./config.toml`.  If you want to put `config.toml` somewhere else (or somewhere specific), you can replace `$(pwd)/config.toml` in the above with an absolute path. (You can do the same with `$(pwd)/data`.)

I’m not familiar with Portainer, so I don’t know how it has you enter the Docker parameters, but if it’s different from the command line invocation, can you, like, take a screenshot of the interface you used (i.e. where you enter the parameters)? The parameter in question should look like one of the following:

* `docker-run -v`
* `docker-run --volume`
* “Volume”, etc.

**Reply by surfjz117** [reply]
First of all thank you for your detailed answer. 

i got it running without portainer using your docker command. Never tried it before cause i was thinking pwd stand for password and than i was to confused to think forward. ...

so now i'm able to run the container using 
```
>  docker run     
> -p 8080:8080    
> -v $(pwd)/data:/app/data/     
> -v $(pwd)/config.toml:/app/config.toml     
> tdeutsch/podsync:latest

```

than i had to remove second autor in example toml and the container  is running. 

In my browser i now get `404 page not found` while opening the url using IP adress or Hostname `http://192.168.1.29:8080/`

(first Error)

In podsync.log i found `time="2022-01-25T08:35:31Z" level=error msg="failed to update feed: https://www.youtube.com/c/ct3003" error="update failed: failed to parse URL: https://www.youtube.com/c/ct3003: unsupported link format"`

i tried another playlist but then my api token quota exceeded. 

So i'll try again later. But 404 is still the next error, i'll have to fix. 

once again : thank you very much for your help!

**Reply by bcarney66** [reply]
The YouTube URL is in the incorrect format, it needs the channel ID in it.   
If the URL isn't in a format like: https://www.youtube.com/channel/UCvUkl9eiDLPtd6SWom413mQ
go to: https://commentpicker.com/youtube-channel-id.php and paste the URL in and it will give you the channel ID.

**Reply by surfjz117** [reply]
Thanks to bcarney66 -- this fixed the issue with channel id

the 404 i fixed by disabling the hostname: 

`#hostname =` in config.toml

and now it`s working. thank you very much!!!!
